src/lua/uri.lua

```diff
@@ -60,6 +60,41 @@ local uri_set_stash = buffer.ffi_stash_new('struct uri_set')
 local uri_set_stash_take = uri_set_stash.take
 local uri_set_stash_put = uri_set_stash.put
 
+-- RFC 3986: Uniform Resource Identifier (URI): Generic Syntax
+-- 2.2. Reserved Characters
+-- https://datatracker.ietf.org/doc/html/rfc3986.html#section-2.2
+local UNRESERVED_RFC3986 = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-._~"
+
+local RFC3986 = {
+    unreserved = UNRESERVED_RFC3986,
+    plus = false,
+}
+
+local FORM_DATA = {
+    unreserved = UNRESERVED_RFC3986,
+    plus = true,
+}
+
+local PATH_SEGMENT = {
+    unreserved = UNRESERVED_RFC3986 .. ":@-._~!$&'()*+,;=",
+    plus = false,
+}
+
+local PATH = {
+    unreserved = UNRESERVED_RFC3986 .. ":@-._~!$&'()*+,;=/",
+    plus = false,
+}
+
+local QUERY = {
+    unreserved = UNRESERVED_RFC3986 .. "/",
+    plus = false,
+}
```

Comment:
AFAIU, here we should escape everything that may be confused with a path segment separator (`/`), a query parameter separators (`=`, `&` and maybe also `?`) and fragment identifier (anchor) separator (`#`). It is just from meaning of the escaping the query parameter's key and value: we need it to don't confuse data with elements of URI structure at decoding. It is likely not a full set, but it looks required.